### PR TITLE
feat(emulator): adds resumable upload API support for multiple chunk and cancelling

### PR DESCRIPTION
**Pull request by jakeisonline**
### Description

This is a follow up to #9047 which was unfortunately reverted in #9107 due to issues in failing integration tests. While it wasn't revealed what the specific issue were, I'm hoping this new approach might avoid those same issues. Given the complexity of resumable uploads, I figured this was an important piece of coverage for the emulator to have.

This PR pays particular attention to the validation of the `Content-Range` header sent by the client, in addition to ensuring the status codes and response bodies (and therefore the underlying logic) coming from the emulator match live GCS as much as possible.

### Approach

* Matching documentation per https://docs.cloud.google.com/storage/docs/performing-resumable-uploads
* Created a set of tests to ensure the correct responses of the API (not included in this PR, as I was unsure what best practises for testing endpoints in this repo are. Happy to include at request, though)
* Created a thin node app which helped me manually test and compare responses in the emulator vs live GCS, including testing edge cases (e.g. malformed `Content-Range`, non-existent upload IDs, etc).
* Using my lived experience working with GCS resumable uploads in a production environment with real end-users

### Scenarios Tested

* PUT
  * single chunk uploads
    * returns 200 + metadata when complete
  * multiple chunk uploads
    * returns 308 + range header when uploading a partial chunk
    * returns 200 + metadata when complete
    * returns 200 when a single chunk uses unknown total (*)
    * returns 308 then 200 when unknown total and the last chunk is not a full 256 KiB multiple
    * returns 308 (not 200) when unknown total and chunk is an exact 256 KiB multiple
  * status checking
    * returns 308 with no Range header when active and no bytes persisted yet
    * returns 308 with Range when active and some bytes are already persisted
    * returns 200 with object metadata when the resumable session is already finished
    * returns * in the Range suffix when bytes */* is used
    * finalises when bytes */total exactly matches bytes received
    * finalises an exact 256 KiB upload via a zero-content bytes */total request
  * error handling
    * returns 400 when content-range is invalid
    * returns 400 when chunk position is invalid
    * returns 400 when chunk size is invalid
    * returns 404 when the upload is not found
    * returns 499 when the upload is cancelled
* DELETE
  * cancels an upload and returns 499
  * returns 404 when upload_id is not found
  * returns 405 when upload_id is missing

### CODE REVIEW

File: src/emulator/storage/apis/gcloud.ts [hunk 1]
```diff
@@ -183,34 +183,139 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
...
+      if (contentLength && contentLength !== "0") {
+        const data = await reqBodyToBuffer(req);
+
+        // Multiple chunk upload
+        if (contentRange) {
+          const parsedRange = parseContentRangeHeader(contentRange);
+
+          if (!parsedRange) {
+            return res.status(400).send("Failed to parse Content-Range header.");
+          }
```
[the remark this answers is not in the document]
Comment (jakeisonline) [reply src/emulator/storage/apis/gcloud.ts]: Resolved per suggestion; reused the already assigned `parsedRange` 🤦

File: src/emulator/storage/apis/gcloud.ts [hunk 2]
```diff
@@ -183,34 +183,139 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
...
+              .send(
+                `Invalid chunk position. The next chunk should start at offset ${upload.size}.`,
+              );
+          }
+
+          const expectedChunkSize = parsedRange.end - parsedRange.start + 1;
+
+          if (data.byteLength !== expectedChunkSize) {
+            const startingOffset = Math.max(parsedRange.start - 1, 0);
+            const message = `Invalid request.  There were ${data.byteLength} byte(s) in the request body.  There should have been ${expectedChunkSize} byte(s) (starting at offset ${startingOffset} and ending at offset ${parsedRange.end}) according to the Content-Range header.`;
```
[the remark this answers is not in the document]
Comment (jakeisonline) [reply src/emulator/storage/apis/gcloud.ts]: Resolved per suggestion.

File: src/emulator/storage/apis/gcloud.ts [hunk 3]
```diff
@@ -462,3 +567,20 @@ function getIncomingFileNameFromRequest(
   const name = query?.name?.toString() || metadata?.name;
   return name?.startsWith("/") ? name.slice(1) : name;
 }
+
+function parseContentRangeHeader(contentRange: string): ContentRange | null {
+  let parsedRange: ContentRange | null = null;
+  const match = /^bytes (\d+)-(\d+)(?:\/(\d+|\*))?$/.exec(contentRange);
```
[the remark this answers is not in the document]
Comment (jakeisonline) [reply src/emulator/storage/apis/gcloud.ts]: Ah, yeah this is certainly my oversight. I'll follow up with a fix, but it requires a little more refactoring that just a regex adjustment.

[the remark this answers is not in the document]
Comment (jakeisonline) [reply src/emulator/storage/apis/gcloud.ts]: Refactored to make sure the case Gemini found is also covered (plus undoing some premature extraction)

**Comment by jakeisonline** [comment]
All Gemini feedback has been resolved. Would it be possible to approve the workflow? I'm particularly interested in the integration tests, which @joehan mentioned were a probably in the previous attempt (#9047). Thanks!

File: src/emulator/storage/apis/gcloud.ts [hunk 4]
```diff
@@ -183,34 +183,157 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
...
       if (err instanceof NotFoundError) {
         return res.sendStatus(404);
+      } else if (err instanceof ForbiddenError) {
+        return res.sendStatus(403);
       } else if (err instanceof UploadNotActiveError) {
         return res.sendStatus(400);
       }
       throw err;
     }
+  });
```
[the remark this answers is not in the document]
Comment (jakeisonline) [reply src/emulator/storage/apis/gcloud.ts:334]: This has now been refactored to reduce the nesting where possible, with some exceptions:

1. `handleStatusCheck` and `handleChunkedUpload` both rely on shared context which includes `uploadService` and `adminStorageLayer` (in addition to the props suggested by Gemini) so these helpers are instead nested within the upload endpoint logic
2. `handleSingleChunkUpload` seemed like premature extraction, it's 3 lines of code.

File: src/emulator/storage/apis/gcloud.ts [hunk 5]
```diff
@@ -183,34 +183,157 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
...
+
+  gcloudStorageAPI.delete("/upload/storage/v1/b/:bucketId/o", async (req, res) => {
+    if (!req.query.upload_id) {
+      return res.sendStatus(405);
```
[the remark this answers is not in the document]
Comment (jakeisonline) [reply src/emulator/storage/apis/gcloud.ts:338]: I agree with this comment, but the live service will respond with a `405` if you supply it with no or invalid params when sending a `DELETE`. Given this is an emulator, I felt it appropriate to follow the live service's responses rather than general REST API best practices.

Happy to be challenged and adjust though.
